### PR TITLE
Add embedded layer tree widget

### LINKED ISSUE #1115: [FEATURE] Layer tree embedded widgets
Original commit: https://github.com/qgis/QGIS/commit/9c7dbb9e0931af6231b0a0314b60ce59decd93e0 by wonder-sk

This allows definition of widgets embedded into layer tree for individual layers in the layer properties dialog (in new Legend tab). The idea is to have a way to quickly access to some actions that are often used with a layer.

The implementation comes with transparency widget, in the future there may be more standard widgets coming, e.g. to setup filtering, selection, style or other stuff. The API allows plugins to register their own widgets, which will be useful for various domain specific plugins to assign custom widgets to layers they manage.

### PR DESCRIPTION
**Pull request by DelazJ**
Fixes #1115
credits to @rduivenvoorde
https://qgis.nl/2019/02/14/about-layer-tree-embedded-widgets-and-have-your-wmts-always-crispy-sharp/?lang=en

### CODE REVIEW

**Comment by DelazJ** [comment]
I'd appreciate a review, as I'm unable to confirm the part concerning the reuse within a plugin

File: docs/pyqgis_developer_cookbook/plugins/snippets.rst [hunk 1]
```diff
@@ -180,6 +189,80 @@ Finally we are adding the imports and modifying the ``__init__`` function:
...
+   provider = LayerStyleWidgetProvider()
+   QgsGui.layerTreeEmbeddedWidgetRegistry().addProvider(provider)
+
+Then from a given layer's :guilabel:`Legend` properties tab, drag the
+``Layer style chooser`` from the :guilabel:`Available widgets` to
+:guilabel:`Used widgets` to enable the widget in the layer tree.
+Embedded widgets are ALWAYS displayed at the top of their associated layer
+node subitems.
+
+If you want to use the widgets from within e.g. a plugin, you can add them
```
Comment (anitagraser) [inline docs/pyqgis_developer_cookbook/plugins/snippets.rst:252]: @wonder-sk Hi Martin, could you please have a look if this is the correct way to use the embedded widgets?

Comment (DelazJ) [reply docs/pyqgis_developer_cookbook/plugins/snippets.rst:252]: Up @wonder-sk ? Thanks.

**Comment by DelazJ** [comment]
I suggest we merge; people know this is possible and if there are any errors in the code, they will probably report.

**Pull request merged**